File: 2023/day06/solution.py
#!/usr/bin/env python3
import sys
from functools import cache

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True
D4 = [(-1, 0), (0, 1), (1, 0), (0, -1)]

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i':
        FILENAME = 'input.txt'
      case '-p1':
        PART1, PART2 = True, False
      case '-p2':
        PART1, PART2 = False, True
      case '-p0':
        PART1, PART2 = False, False
      case _:
        raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  lines = [x.split() for x in f.read().splitlines()]
races = [{
  'time': int(x),
  'record': int(y),
} for x, y in zip(lines[0][1:], lines[1][1:])]
if DEBUG > 1: print(f'{races=}')


def part1():
  # i think the question specifically wants a brute forced answer :(
  total = 1
  for race in races:
    race_total = 0
    for i in range(0, race['time'] + 1):
      speed = i
      time = race['time'] - i
      dist = speed * time
      win = dist > race['record']
      if DEBUG > 1: print(f'{race=} {i=} {speed=} {time=} {dist=} {win=}')
      if win: race_total += 1
    total *= race_total

  print(f'part 1: {total=}')


def part2():
  race_time = int(''.join(str(x['time']) for x in races))
  record = int(''.join(str(x['record']) for x in races))
  if DEBUG > 1: print(f'{race_time=} {record=}')
  # binary search the lowest press time to win
  lower = 0
  upper = race_time
  while upper > lower:
    mid = (lower + upper) // 2
    win = mid * (race_time - mid) > record
    if win:
      upper = mid
    else:
      lower = mid + 1
  print(f'{upper=} {lower=} {win=}')
  win_lower = lower

  # and again for the highest
  lower = win_lower + 1
  upper = race_time
  while upper > lower:
    mid = (lower + upper) // 2
    win = mid * (race_time - mid) > record
    if win:
      lower = mid + 1
    else:
      upper = mid
  print(f'{upper=} {lower=} {win=}')
  win_upper = lower

  total = win_upper - win_lower
  print(f'part 2: {total=}')


if PART1: part1()
if PART2: part2()

File: 2023/day06/solution.ts
#!/usr/bin/env -S deno --allow-read

const Maths = Math;

const DEBUG = Deno.args.reduce((acc, item) => (item.startsWith('-d') ? Number(item.slice(2) || '1') : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) => (item == '-i' ? 'input.txt' : item.startsWith('-e') ? `example${item.slice(2)}.txt` : acc),
  'example.txt'
);
const [PART1, PART2] = Deno.args.reduce(
  (acc, item) => (item == '-p0' ? [false, false] : item == '-p1' ? [true, false] : item == '-p2' ? [false, true] : acc),
  [true, true]
);

console.log({ FILENAME, DEBUG, PART1, PART2 });

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (DEBUG >= level) console.debug(...data);
};

const input = await Deno.readTextFile(FILENAME);

interface Race {
  time: number;
  distance: number;
}

const part1 = () => {
  const tokens = input
    .split('\n')
    .filter((line) => line.trim())
    .map((line) =>
      line
        .split(/\s+/)
        .slice(1)
        .map((token) => Number(token))
    );
  const races: Race[] = tokens[0].map((token, i) => ({ time: token, distance: tokens[1][i] }));

  const solutionCounts = races.map((race) => {
    debug(2, { race });
    let solutions = 0; //FIXME: let is a moral failing
    for (let speed = 0; speed <= race.time; ++speed) {
      if (speed * (race.time - speed) > race.distance) ++solutions;
    }
    debug(2, { race, solutions });
    return solutions;
  });

  debug(1, solutionCounts);
  const product = solutionCounts.reduce((acc, item) => acc * item, 1);
  console.log('part 1:', product);
};

const part2 = () => {
  const tokens = input
    .split('\n')
    .filter((line) => line.trim())
    .map((line) => Number(line.split(/\s+/).slice(1).join('')));
  const race: Race = { time: tokens[0], distance: tokens[1] };

  const binarySearch = (low: number, high: number, test: (mid: number) => boolean): number => {
    if (low == high) {
      debug(1, { solution: low });
      return low;
    }
    const mid = Maths.floor((low + high) / 2);
    const result = test(mid);
    debug(2, { high, low, mid, result });
    return binarySearch(result ? low : mid + 1, result ? mid : high, test);
  };

  const lowestWin = binarySearch(0, race.time, (mid) => mid * (race.time - mid) > race.distance);
  const highestWin = binarySearch(lowestWin, race.time, (mid) => mid * (race.time - mid) <= race.distance);
  const solutions = highestWin - lowestWin;
  console.log('part 2:', solutions);
};

if (PART1) part1();
if (PART2) part2();
